### PR TITLE
NDRS-473: Optimize bytes/array serialization without nightly

### PR DESCRIPTION
**Pull request by mpapierski**
Ref: https://casperlabs.atlassian.net/browse/NDRS-473

One step forward for getting stable support. Removes all default impls, and where applicable `Vec<u8>`/`[u8; N]` serialiation/deserialization is done through newtypes wrapper where it is applied at type level (i.e. replaced `Vec<u8>` with `Bytes`. Array serialization/deserialization is limited only to u8 and this also reflects in `CLType` where `FixedList(type, size)` is replaced with `ByteArray(size)`.

Benchmarks https://gist.github.com/mpapierski/226df42d4705dcb95e5094ba9b910bad against master

### CODE REVIEW

File: execution_engine/src/core/runtime/mod.rs [hunk 1]
```diff
@@ -183,403 +183,7 @@ fn extract_urefs(cl_value: &CLValue) -> Result<Vec<URef>, Error> {
...
-                Ok(arr.to_vec())
-            }
-            CLType::Key => {
-                let arr: [Key; 512] = cl_value.to_owned().into_t()?;
-                Ok(arr.iter().filter_map(Key::as_uref).cloned().collect())
-            }
-            _ => Ok(vec![]),
-        },
-        CLType::FixedList(_ty, _) => Ok(vec![]),
+        CLType::ByteArray(_) => Ok(vec![]),
```
Comment (dwerner) [inline execution_engine/src/core/runtime/mod.rs:186]: :+1:

File: execution_engine/src/storage/trie/mod.rs [hunk 2]
```diff
@@ -129,22 +133,44 @@ impl Default for PointerBlock {
...
+        self.0.iter().map(ToBytes::serialized_length).sum()
     }
 }
 
 impl FromBytes for PointerBlock {
-    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
-        FromBytes::from_bytes(bytes).map(|(arr, rem)| (PointerBlock(arr), rem))
+    fn from_bytes(mut bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let pointer_block_array = {
+            let mut result: MaybeUninit<PointerBlockArray> = MaybeUninit::uninit();
```
Comment (dwerner) [inline execution_engine/src/storage/trie/mod.rs:151]: I'm having a hard time understanding this at first glance - would it be possible to add a comment that describes the justification for using `MaybeUninit`?

File: types/src/gens.rs [hunk 3]
```diff
@@ -203,7 +200,7 @@ pub fn cl_value_arb() -> impl Strategy<Value = CLValue> {
         ".*".prop_map(|x: String| CLValue::from_t(x).expect("should create CLValue")),
         option::of(any::<u64>()).prop_map(|x| CLValue::from_t(x).expect("should create CLValue")),
         vec(uref_arb(), 0..100).prop_map(|x| CLValue::from_t(x).expect("should create CLValue")),
-        [any::<u64>(); 32].prop_map(|x| CLValue::from_t(x).expect("should create CLValue")),
+        // [any::<u64>(); 32].prop_map(|x| CLValue::from_t(x).expect("should create CLValue")),
```
Comment (dwerner) [inline types/src/gens.rs:203]: If it's intended to leave this commented out, perhaps a comment as to why?

**Review by dwerner — APPROVED** [review]
Nice job, cleans up a lot of code, left a couple of nitpicks about commenting some areas that weren't clear to me. Do you have any criterion/bench results showing the speedup?

**Review by henrytill — COMMENTED** [review]
Code itself looks good, a few nits. 

However, my benchmark results do show some not-insignificant regressions, along with some improvements.  A mixed bag...
https://gist.github.com/henrytill/03ff03322c8aa26c017f556f65d83857

Do we still have the transfer benchmarks?  They might be helpful in helping us assess the real-world performance impact of these changes.  I was under the impression they would be here, but don't seem to be included anymore.

cc @EdHastingsCasperLabs

**Comment by henrytill** [comment]
> Do we still have the transfer benchmarks? They might be helpful in helping us assess the real-world performance impact of these changes. I was under the impression they would be here, but don't seem to be included anymore.

Figured it out and got some transfer bench runs:
https://gist.github.com/henrytill/03ff03322c8aa26c017f556f65d83857#file-transfer_benchmark-txt

:+1:

**Review by henrytill — APPROVED** [review]

**Review by Fraser999 — CHANGES_REQUESTED** [review]
The following comments on the previous PR have not been addressed:
* https://github.com/CasperLabs/casper-node/pull/526#discussion_r522955722
* https://github.com/CasperLabs/casper-node/pull/526#discussion_r522920861
* https://github.com/CasperLabs/casper-node/pull/526#discussion_r522975422
* https://github.com/CasperLabs/casper-node/pull/526#discussion_r522979787

**Review by Fraser999 — APPROVED** [review]

**Pull request merged**